File: src/main/scala/lang/Token.scala
package cql.lang

import io.circe.Encoder
import io.circe.Json
import io.circe.syntax._

enum TokenType:
  // Single-character tokens.
  case PLUS, COLON, AT, LEFT_BRACKET, RIGHT_BRACKET,

    // Literals.
    STRING, NUMBER, QUERY_OUTPUT_MODIFIER_KEY, CHIP_KEY, CHIP_VALUE,

    // Keywords.
    AND, OR, EOF

case class Token(
    tokenType: TokenType,
    lexeme: String,
    literal: Option[String],
    start: Int,
    end: Int
):
  override def toString = s"${tokenType} ${lexeme} ${literal} ${start}-${end}"

object Token:
  val reservedWords = Map(
    "AND" -> TokenType.AND,
    "OR" -> TokenType.OR
  )

  implicit val tokenEncoder: Encoder[Token] = Encoder.instance { token =>
    Json.obj(
      "type" -> "Token".asJson,
      "tokenType" -> token.tokenType.toString.asJson,
      "lexeme" -> token.lexeme.asJson,
      "start" -> token.start.asJson,
      "end" -> token.end.asJson,
      "literal" -> token.literal.asJson
    )
  }

File: client/src/lang/token.ts
export type TokenType = keyof typeof TokenType;

export const TokenType = {
  // Single-character tokens.
  LEFT_BRACKET: "LEFT_BRACKET",
  RIGHT_BRACKET: "RIGHT_BRACKET",

  // Literals.
  STRING: "STRING",
  CHIP_KEY_POSITIVE: "CHIP_KEY_POSITIVE",
  CHIP_KEY_NEGATIVE: "CHIP_KEY_NEGATIVE",
  CHIP_VALUE: "CHIP_VALUE",

  // Keywords.
  AND: "AND",
  OR: "OR",
  EOF: "EOF",
} as const;

export const isChipKey = (tokenType?: TokenType) =>
  tokenType &&
  (tokenType === TokenType.CHIP_KEY_POSITIVE ||
    tokenType === TokenType.CHIP_KEY_NEGATIVE);

export class Token {
  public static reservedWordMap = {
    AND: TokenType.AND,
    OR: TokenType.OR,
  } as const;

  public static reservedWordStrs = Object.keys(this.reservedWordMap);

  constructor(
    public tokenType: TokenType,
    public lexeme: string,
    public literal: string | undefined,
    public start: number,
    public end: number
  ) {}
  public toString() {
    return `${this.tokenType} ${this.lexeme} ${this.literal} ${this.start}-${this.end}`;
  }
}
